js/misc/missionGameService.js

```diff
@@ -0,0 +1,267 @@
+ * This is used by MissionChatboxTextService to ensure that
+ * all asynchronous tasks are completed before calling
+ * the completion callback. The callback is called with
+ * the key that triggered the callback.
+ */
+function callWhenComplete(completion, key, callback) {
+    completion[key] = true;
+
+    /* This is a shorthand way of implementing Array.all
+     * using De Morgan's laws and Array.some */
+    const complete = !Object.keys(completion).some(function(key) {
+        return !completion[key];
+    });
+
+    if (complete && callback) {
+        callback.apply(this, [key].concat(Array.from(arguments)));
+    }
+}
+
+const MissionChatboxTextService = new Lang.Class({
+    Name: 'MissionChatboxTextService',
+    Extends: GLib.GObject,
+
+    _init: function(props) {
+        this.parent(props);
+
+        /* Null-initialise service for now, but we'll set it later */
+        this._service = null;
+
+        /* These properties make up the lesson's content. _lessonInfo refers
+         * to all of the text we need to display about the broader "lesson"
+         * and _taskInfo is all the text that we need to display about the
+         * currently active task.
+         *
+         * The state of this service moves through each each task until
+         * it reaches the end, at which point it displays a message
+         * and sets itself to a "done" state.
+         */
+        this._openedForTheFirstTime = false;
+        this._introLesson = null;
+        this._currentTask = null;
+
+        const name = "com.endlessm.Showmehow.Service";
+        const path = "/com/endlessm/Showmehow/Service";
+
+        /* Connect to the service and refresh the content once we have a connection */
+        Showmehow.ServiceProxy.new_for_bus(Gio.BusType.SESSION, 0, name, path, null,
+                                           Lang.bind(this, function(source, result) {
+            this._service = Showmehow.ServiceProxy.new_for_bus_finish(result);
+            this._service.connect("lessons-changed", Lang.bind(this, function(proxy) {
+                /* When the underlying lessons change, we need to reset the entire
+                 * state of this component and start from the beginning, including
+                 * showing any warnings.
+                 *
+                 * Get the intro lesson again then reset all the state back
+                 * to its initial point.
+                 */
+                this._refreshContent(function() {
+                    this._openedForTheFirstTime = false;
+                    this._currentTask = null;
+                });
+            }));
+            this._refreshContent();
+        }));
+    },
+    commenceIntroLesson: function() {
+        /* If possible in the current state, commence the intro lesson.
+         *
+         * That depends on us having not been opened for the first
+         * time and the content being loaded. */
+        if (!this._openedForTheFirstTime && this._introLesson) {
```

Comment:
Ah, right, having a look at the context, I see your point. I'll return early. Thanks.